packages/belos/tpetra/src/BelosTpetraOperator.hpp

```diff
@@ -0,0 +1,200 @@
+#include "BelosStatusTest.hpp"
+#include "BelosOutputManager.hpp"
+#include "BelosTpetraAdapter.hpp"
+#include "BelosSolverFactory.hpp"
+
+#include "Teuchos_ParameterList.hpp"
+
+/*! \class Belos::TpetraOperator
+    \brief This class provides an interface to the Tpetra::Operator class, so Belos can be 
+    integrated into other codes as an abstract operator.
+*/
+
+namespace Belos {
+
+///////////////////////////////////////////////////////////////
+//-------- class BelosTpetraOperator --------------------
+//
+// This class will allow a Belos solver to be called as an Tpetra_Operator.
+// Thus, it can use itself as a preconditioner if need be.  It can
+// also be used as the inner iteration of Anasazi :)
+//
+///////////////////////////////////////////////////////////////
+template<class Scalar = ::Tpetra::Details::DefaultTypes::scalar_type,
+         class LocalOrdinal = ::Tpetra::Details::DefaultTypes::local_ordinal_type,
+         class GlobalOrdinal = ::Tpetra::Details::DefaultTypes::global_ordinal_type,
+         class Node = ::Tpetra::Details::DefaultTypes::node_type>
+class TpetraOperator : public virtual Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> {
+public:
+  //! @name Typedefs
+  //@{
+  //! The Tpetra Multivector 
+  typedef Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_MultiVector;
+  //! The Tpetra Operator
+  typedef Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_Operator;
+  //! The Tpetra Map
+  typedef Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> Tpetra_Map;
+
+  //! @name Constructor / Destructor
+  //@{ 
+  
+  //! Constructor
+  TpetraOperator( const Teuchos::RCP<LinearProblem<Scalar,Tpetra_MultiVector,Tpetra_Operator> >& lp, 
+		  const Teuchos::RCP<Teuchos::ParameterList>& plist,
+                  bool initSolnVec = false )  
+  : lp_(lp), 
+    plist_(plist),
+    initSolnVec_(initSolnVec)
+  {
+  // Get the solver's name from the parameter list, use block Gmres by default.
+  std::string solver = plist_->get("Solver", "BlockGmres");
+
+  // Create a label for this Tpetra_Operator.
+  std::string solver_name = "Belos " + solver + " Solver";
+ 
+  // Copy std::string to character array.  
+  // Not using conversion routine copy() because it's not supported by RW on Janus. (HKT 11/13/2003) 
+  Solver.resize(solver_name.length()+1);
+  for (int i=0; i<(int)solver_name.length(); i++) {
+    Solver[i] = solver_name[i];
+  }   
+  Solver[solver_name.length()] = 0;
+
+  //  
+  // Create solver and solve problem.  This is inefficient, an instance of the solver should
+  // exist already and just be reset with a new RHS.
+  //  
+  if (solver == "BlockGmres") {
+    solver = "Block Gmres";
+  }   
+  else if (solver == "PseudoBlockGmres") {
+    solver = "Pseudo Block Gmres";
+  }
+  else if (solver == "BlockCG") {
+    solver = "Block CG";
+  }
+  else if (solver == "PseudoBlockCG") {
+    solver = "Pseudo Block CG";
+  }
+
+  plist_->remove( "Solver" );
```

Comment:
Why remove the parameter?

Reply:
?? @hkthorn

Reply:
See below.

Reply:
It's probably because solvers don't know what this "Solver" parameter is.  That suggests we should make the solver name separate from the ParameterList.

Reply:
Moving it to be a parameter to the constructor.